examples/vision/nnclr.py

```diff
@@ -0,0 +1,642 @@
+verification/
+identification](https://openaccess.thecvf.com/content_cvpr_2015/html/Schroff_FaceNet_A_Uni
+fied_2015_CVPR_paper.html). These methods learn a latent space that draws positive
+samples together while pushing apart negative samples.
+
+### NNCLR
+
+In this example we will be implementing NNCLR as proposed in the paper [With a Little
+Help from My Friends:
+Nearest-Neighbor Contrastive Learning of Visual
+Representations](https://arxiv.org/abs/2104.14548), by Google Research and DeepMind
+which:
+- learns  self-supervised representations that go beyond single instance positives which
+allows to learn better features that are invariant to different viewpoints, deformations,
+and even intra-class variations
+- Clustering based methods offer a great approach to go beyond single instance positives,
+but assuming the entire cluster  to be positives could hurt performance due to early
+over-generalization. Instead NNCLR uses nearest neighbors in the learned representation
+space as positives
+- increases the performance of existing contrastive learning methods like
+[SimCLR](https://arxiv.org/abs/2002.05709)([Keras
+Example](https://keras.io/examples/vision/semisupervised_simclr))
+- reduces the reliance of self-supervised methods on data augmentation strategies
+
+Here is a great visualization by the paper authors showing how NNCLR builds on ideas from
+SimCLR:
+
+![](https://i.imgur.com/p2DbZJJ.png)
+
+We can see that SimCLR uses two views of the same image as the positive pair. These two
+views, which are produced using random data augmentations, are fed through an encoder to
+obtain the positive embedding pair, we end up using two augmentations. NNCLR instead
+keeps a _support set_ of embeddings representing the full data distribution, and forms
+the positive pairs using nearest-neighbours. A support set is used as memory during
+training; similar to a queue (i.e. first-in-first-out) like in
+[MoCo](https://arxiv.org/abs/1911.05722).
+
+This example requires TensorFlow 2.4 or higher as well as `tensorflow_datasets` which can
+be installed with this command:
+"""
+
+"""shell
+pip install tensorflow-datasets
+"""
+
+"""
+## Setup
+"""
+
+import random
+
+import matplotlib.pyplot as plt
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow import keras
+from tensorflow.keras import layers
+
+"""
+## Hyperparameters
+
+A greater `queue_size` most likely means better performance as also shown in the original
+paper but introduces a huge computational overhead. The authors show that the best
+results of NNCLR are achieved with a queue size of 98,304 (the largest `queue_size` they
+experimented on). We here use 10,000 to show a working example.
+"""
+
+AUTOTUNE = tf.data.experimental.AUTOTUNE
+shuffle_buffer = 5000
+# The below two values are taken from https://www.tensorflow.org/datasets/catalog/stl10
+labelled_train_images = 5000
+unlabelled_images = 100000
+
+temperature = (0.1,)
```

Comment:
This should probably be a float, not a tuple

Reply:
Yes, in fact, a tuple would not even work, right? The brackets seem to have been added by black due to a trailing comma. Fixed this.